### DISCUSSION TITLE
Setting up timelapse

### DISCUSSION THREAD

**Discussion by PitbullRaven** [nlef/moonraker-telegram-bot#162] [Q&A] [answered]
Sorry for being a daft prick, but I just managed to setup the bot, though I cannot figure out how to set up time-lapse. I have done the optional step 5 during install.Should I be changing something in that script?

**Comment by aka13-404** [comment, marked as answer]
Hello my friend, don't worry, you are no a daft prick. 

Here is how you would proceed:
1) The bot works "like klipper", if you need some functionality, you add a corresponding section with the needed parameters
2) All the stuff is documented on the wiki, third tab from the right.
3) You open the sample config section on the wiki, and scroll down to the timelapse section. There you can read, what each parameter does and pick and choose the ones you need. 
4) Don't forget to restart the bot when you are done, so that the changes are applied.

Let me know how it works out for you.

**Reply by PitbullRaven** [reply]
Thanks for the reply. I managed to get my first time lapse going.

**Reply by PitbullRaven** [reply]
By the way... Is there any way of getting the bot to capture a last frame after the print ends and the head moves back to the parked position? Most time-lapses end with the head still over the print, thus obscuring part of the finished product.

https://user-images.githubusercontent.com/14200375/171631129-9061c244-cada-4563-bde8-728e83d467f8.mp4

**Comment by aka13-404** [comment]
Sure, you can shoot a photo at any time in addition to the automated options.
https://github.com/nlef/moonraker-telegram-bot/wiki/Interacting-with-klipper#controlling-timelapse-parameters-via-gcode

**Reply by mrmhger** [reply]
hi, could you please help me to understand how manual mode works? i didnt find any info besides you're linking at, but i dont really understand that. "respond" commands i.E. how to activate them? how to use? or at least where can i find some info about it? my google ist awesome. thx a lot.

**Reply by aka13-404** [reply]
It's a klipper command to write stuff into console. By sending it with a prefix you instruct the bot to react to stuff written in the console.
Refer to klipper documentation: https://github.com/Klipper3d/klipper/blob/master/docs/G-Codes.md#respond

**Reply by PitbullRaven** [reply]
I've been trying to get my brain to understand how I can move the extruder out of the way before the last photo, but nothing occurs to me. It seems that I cannot do that in auto mode. Meaning, I'd have to do it all via the G-Code, right? Or is there an easier way? Right now, I have the software creating the timelapse automagically (and running stable so far). I really didn't want to mess it up too much.

**Reply by aka13-404** [reply]
No, not at all, just put the automating lapsing to a pause, move the head, take the picture, run timelapse stop, run timelapse create.

https://github.com/nlef/moonraker-telegram-bot/wiki/Interacting-with-klipper#controlling-timelapse-parameters-via-gcode

Works the same way the manual photos do.

**Reply by PitbullRaven** [reply]
So pause at the end of the print using GCODE? I suppose that entails adding the correct GCODE commands to the end script? Sorry for so many questions, but this is a bit more involved than anything I have attempted on Klipper so far.

**Reply by aka13-404** [reply]
You must have some sort of a print-ending gcode, right? Where you say, "it's done, turn the heaters off"? Insert a automated-lapse pause there, move your head to where you want it, take picture, stop lapse, build lapse.

[gcode_macro END_PRINT]
gcode:
RESPOND PREFIX=timelapse MSG=pause
#move head away
RESPOND PREFIX=timelapse MSG=photo
RESPOND PREFIX=timelapse MSG=stop
#whatever else you might do from here on, disable heaters yada yada
RESPOND PREFIX=timelapse MSG=create

**Reply by PitbullRaven** [reply]
This is what I assumed, and I was assuming I'd have to place this either in the SuperSlicer End G-Code or the printer.cfg file on my printer. But the code doesn't work in SuperSlicer and my printer.cfg does not have an END section in it.

Edit:
Never mind. I think I figured it out. I had to add the code to the printer.cfg and then alter it a bit for the end move, etc... Testing it now.

**Reply by PitbullRaven** [reply]
So it didn't work. Here is my entire [printer.cfg](https://pastebin.com/wRJVGj5d). I think that the problem might stem from line 154, maybe? Can I rename the macro to anything else and then just change the end g-code section in SuperSlicer to invoke that macro?

**Reply by aka13-404** [reply]
Try the klipper discord or any other chat of your preference on help with macros and g-codes.
This is the wrong place for that kind of issues.

**Reply by PitbullRaven** [reply]
@aka13-404 , I want to thank you for the help you have already given me. Much appreciated.

**Reply by PitbullRaven** [reply]
So I had to tweak it a bit, but thanks to your initial indications I was able to arrive at this code:
[gcode_macro LAST_PIC]
gcode:
  RESPOND PREFIX=timelapse MSG=pause
  # Raise nozzle by 10mm
  G91
  G1 Z10 F3000
    # Move nozzle away from print while retracting
  G90
  G1 X250 Y500 E-3 F300
  # Disable steppers
  #M84
  RESPOND PREFIX=timelapse MSG=photo
  RESPOND PREFIX=timelapse MSG=stop
  #whatever else you might do from here on, disable heaters yada yada
  RESPOND PREFIX=timelapse MSG=create
  END_PRINT

The only thing missing is that it won't upload the finished timelapse to telegram, and I can't see a way to do it in manual mode. Is this correct?

**Reply by aka13-404** [reply]
The "create" message should build and upload the lapse. The code I sent you before, is the one I use for uploading. I will doublecheck that it works with semi-automatic mode as well.

**Reply by PitbullRaven** [reply]
I think I may know what is happening. This is the macro as I have it:
```
[gcode_macro LAST_PIC]
gcode:
  RESPOND PREFIX=timelapse MSG=pause
  # Raise nozzle by 10mm
  G91
  G1 Z10 F3000
    # Move nozzle away from print while retracting
  G90
  G1 X250 Y500 F20000
  # Disable steppers
  #M84
  RESPOND PREFIX=timelapse MSG=photo
  RESPOND PREFIX=timelapse MSG=stop
  #whatever else you might do from here on, disable heaters yada yada
  RESPOND PREFIX=timelapse MSG=create
  END_PRINT
```

In SuperSlicer end code I have the LAST_PIC being called followed by END_PRINT.

When the print ends, it seems to error out. After moving to the position set by G1, it says it must home all axis. This cannot happen because there are now parts on the plate and collisions would be inevitable.

```
02:50:57
Must home axis first: 246.837 495.000 88.920 [1794.168]
02:50:57
Must home axis first: 246.837 495.000 88.920 [1794.168]
02:50:57
Must home axis first: 246.837 495.000 88.920 [1794.168]
02:50:57
echo: Cleaning up...
02:50:57
echo: Done :)
02:50:55
echo: Cleaning up...
02:50:55
timelapse create
02:50:55
timelapse stop
02:50:55
timelapse photo
02:50:54
timelapse pause
02:33:40
echo: Priming nozzle with prime line..
02:32:51
Bed Mesh state has been saved to profile [default]
for the current session. The SAVE_CONFIG command will
```

**Reply by aka13-404** [reply]
I have checked, everything works as expected.
As I have stated before, this repo is the wrong place to learn writing macros.

**Reply by PitbullRaven** [reply]
@aka13-404 Thanks to your pointers and some persistence on my part, I managed to figure out where I was going wrong. I'll leave the macro here, if you don't mind, in case someone finds it useful for the future.

In SuperSlicer I added the LATS_PIC macro trigger AFTER the END_PRINT macro. This way, the END_PRINT macro parks the head away from the part and then the LAST_PIC macro can just take the shot.
In my printer.cfg, the LAST_PIC macro is:
```
[gcode_macro LAST_PIC]
gcode:
  RESPOND PREFIX=timelapse MSG=pause
#Wait 5 seconds before photo
  G4 P5000
  RESPOND PREFIX=timelapse MSG=photo
  RESPOND PREFIX=timelapse MSG=stop
  #whatever else you might do from here on, disable heaters yada yada
  RESPOND PREFIX=timelapse MSG=create
```

One additional question from my side: is it possible to have the first picture also repeat for X seconds before starting the timelapse playback? Right now, I have managed to get the bot to take a first picture still at the edge of the build plate, then it does its magic, then at the end runs this macro, moving the head away before taking the final picture and creating everything. The last picture is repeated for 5 seconds in my case. Was wondering if we can do the same for the first one?
